[k8s] SkyServe on Kubernetes

romilbhardwaj opened the pull request:
Adds SkyServe support on Kubernetes. Replicas and controller can now be run on Kubernetes. Supersedes #3109.

Changes in this PR:
* Support for service accounts #3277
* Docs on minimal permissions #2806
* Refactor `get_endpoints` to backend_utils
* Use `get_endpoints` API in serve controller and LB
* Optimizations to reduce dependencies installed on the controller based on enabled clouds

Tested:
- [x] Code formatting: `bash format.sh`
- [x] HTTP Serve Example on GKE with ingress
- [x] HTTP Serve Example on GKE with LoadBalancer 
- [x] LLM Example on GKE with ingress
- [x] HTTP Serve Example on EKS with ingress
- [x] HTTP Serve Example on EKS with LoadBalancer  
- [x] Onprem testing through design partner
- [x] All smoke tests, including serve tests now enabled for k8s (in progress)

Michaelvll commented on sky/provision/kubernetes/utils.py:
```diff
@@ -509,20 +510,92 @@ def check_credentials(timeout: int = kubernetes.API_TIMEOUT) -> \
+        gpu_msg = str(e)
+    if exec_msg and gpu_msg:
+        return True, f'{gpu_msg}\n    Additionally, {exec_msg}'
+    elif gpu_msg or exec_msg:
+        return True, gpu_msg or exec_msg
```
It might be better to expand this to two `elif` according to the google style guide: https://google.github.io/styleguide/pyguide.html#2144-decision

Michaelvll commented on sky/serve/constants.py:
```diff
@@ -69,7 +69,7 @@
 CONTROLLER_PORT_START = 20001
 LOAD_BALANCER_PORT_START = 30001
-LOAD_BALANCER_PORT_RANGE = '30001-30100'
+LOAD_BALANCER_PORT_RANGE = '30001-30020'
```
Any reason we reduce the range of the LB ports?

romilbhardwaj replied:
Opening many ports in LB mode can be expensive on hosted Kubernetes (GKE/EKS) can be expensive since each will create a load balancer service. On on-prem k8s it may exhaust list of public IPs available on the cluster. Also IIRC creating more than 100 services on GKE requires some quota extension..

Michaelvll commented on sky/serve/controller.py:
```diff
@@ -152,7 +152,7 @@ def configure_logger():
                     f'http://localhost:{self._port}')
 
-        uvicorn.run(self._app, host='localhost', port=self._port)
+        uvicorn.run(self._app, host='0.0.0.0', port=self._port)
```
It would be nice to add a comment here for why `0.0.0.0` instead of `localhost` is used.

Michaelvll commented on sky/clouds/kubernetes.py:
```diff
@@ -45,6 +46,8 @@ class Kubernetes(clouds.Cloud):
     # TODO(romilb): Make the timeout configurable.
     TIMEOUT = 10
 
+    _SUPPORTS_SERVICE_ACCOUNT_ON_REMOTE = True
```
Is this needed?

romilbhardwaj replied:
Yes, it's used by `cloud.supports_service_account_on_remote()` to tell whether to upload local cloud credential files

Michaelvll commented on sky/provision/__init__.py:
```diff
@@ -141,10 +141,15 @@ def query_ports(
+
     Returns a dict with port as the key and a list of common.Endpoint.
     """
     raise NotImplementedError
```
How about use the implementation here to let the API to use the default implementation in this file if it is not implemented in the specific provider?

https://github.com/skypilot-org/skypilot/blob/421e62402ce51576a5668ce139234a26c52f7475/sky/provision/__init__.py#L29-L50

romilbhardwaj replied:
Hmm, not sure if I fully follow - do you have an example in mind? Do you mean:
1. Calling `common.query_ports_passthrough(ports, head_ip)` in query_ports here
2. Modify `_route_to_cloud_impl` to route only if underlying provider implements the method?

romilbhardwaj replied:
Good idea - switched the implementation.

Michaelvll commented on sky/serve/serve_utils.py:
```diff
@@ -808,10 +815,11 @@ def _format_replica_table(replica_records: List[Dict[str, Any]],
         replica_id = record['replica_id']
         version = (record['version'] if 'version' in record else '-')
-        replica_ip = '-'
+        replica_ip = endpoint if endpoint else '-'
```
How about removing this `replica_ip` and using the `endpoint` (rename to `replica_endpoint`) above below?

romilbhardwaj replied:
`record['endpoint']` can be `None` if the handle did not exist or the port was not yet exposed, so I had split the logic into [getting the endpoint](https://github.com/skypilot-org/skypilot/pull/3377#discussion_r1545791568), and then changing it to `-` if it was `None` (could be `None` from the payload or the default value set above).

Another way to do this could be:
```
endpoint = record.get('endpoint', None) if record.get('endpoint', None) else '-' 
```
but that seemed a little verbose

romilbhardwaj replied:
I have renamed `replica_ip` to `replica_endpoint` if that helps

Michaelvll commented on sky/serve/serve_utils.py:
```diff
@@ -808,10 +815,11 @@ def _format_replica_table(replica_records: List[Dict[str, Any]],
         replica_records = replica_records[:_REPLICA_TRUNC_NUM]
 
     for record in replica_records:
+        endpoint = record.get('endpoint', None)
```
```suggestion
        endpoint = record.get('endpoint', '-')
```

Michaelvll replied:
Just a kind bump for this nit.

Michaelvll commented on sky/utils/common_utils.py:
```diff
@@ -61,11 +61,17 @@ def get_usage_run_id() -> str:
 
 
-def get_user_hash() -> str:
+def get_user_hash(force_fresh_hash: bool = False) -> str:
```
any reason for adding this `force_fresh_hash`? It seems we don't use it elsewhere.

romilbhardwaj replied:
It is used in `kubernetes.ssh_key_secret_field_name` to override the user_hash set for the controller through envvars

Michaelvll commented on sky/utils/controller_utils.py:
```diff
@@ -139,34 +139,54 @@ def from_name(cls, name: Optional[str]) -> Optional['Controllers']:
             'azure-mgmt-network > /dev/null 2>&1')
+    # Kubernetes
+    if clouds.cloud_in_list(clouds.Kubernetes(), enabled_clouds):
+        commands.append(
+            # Install k8s + skypilot dependencies
+            'sudo bash -c "if '
+            '! command -v curl &> /dev/null || '
+            '! command -v socat &> /dev/null || '
+            '! command -v netcat &> /dev/null; '
+            'then apt update && apt install curl socat netcat -y; '
+            'fi" && '
+            # Install kubectl
+            '(command -v kubectl &>/dev/null || '
+            '(curl -LO "https://dl.k8s.io/release/$(curl -L -s '
+            'https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl" && '
+            'sudo install -o root -g root -m 0755 '
+            'kubectl /usr/local/bin/kubectl)) && ')
```
Should we consider adding these dependencies in our default kubernetes image?

romilbhardwaj replied:
Tracked in https://github.com/skypilot-org/skypilot/issues/3435

Michaelvll reviewed (commented):
Thanks for adding the support for SkySeve on k8s @romilbhardwaj! This is awesome. Left several comments.

romilbhardwaj commented on docs/source/cloud-setup/cloud-permissions/kubernetes.rst:
```diff
@@ -0,0 +1,233 @@
+When running inside your Kubernetes cluster (e.g., as a Spot controller or Serve controller),
+SkyPilot can operate using either of the following three authentication methods:
+
+1. **Using your local kubeconfig file**: In this case, SkyPilot will
+   copy your local ``~/.kube/config`` file to the controller pod and use it for
+   authentication. This is the default method when running inside the cluster,
+   and no additional configuration is required.
+
+   .. note::
+
+       If your cluster uses exec based authentication in your ``~/.kube/config`` file
+       (e.g., GKE uses exec auth by default), SkyPilot may not be able to authenticate using this method. In this case,
+       consider using the service account methods below.
```
Hmm the Roles listed in this doc should apply to both, the user account or service account configured in kubeconfig and the service account that must be created for SkyPilot controllers.

Added a small note - `These roles must apply to both the user account configured in the kubeconfig file and the service account used by SkyPilot (if configured).` Does that help?

romilbhardwaj commented on docs/source/cloud-setup/cloud-permissions/kubernetes.rst:
```diff
@@ -0,0 +1,233 @@
+
+.. _k8s-permissions:
+
+Permissions required by SkyPilot
```
These permissions apply to both the user account and the service account (if configured), so not using the word `Service Account`.

These are also not the minimal permissions since you could technically specify the exact resources (e.g., pod, service, ingress, secrets, serviceaccounts) instead of using `resources: ["*"]`, but that would be long-ish specification 😅

romilbhardwaj commented on sky/clouds/cloud.py:
```diff
@@ -41,6 +41,7 @@ class CloudImplementationFeatures(enum.Enum):
     SPOT_INSTANCE = 'spot_instance'
     CUSTOM_DISK_TIER = 'custom_disk_tier'
     OPEN_PORTS = 'open_ports'
+    HOST_CONTROLLERS = 'host_controllers'
```
Added in #3407

romilbhardwaj commented on sky/backends/backend_utils.py:
```diff
@@ -2681,3 +2689,101 @@ def check_stale_runtime_on_remote(returncode: int, stderr: str,
+                                 'on cluster '
+                                 f'{cluster_record["name"]!r}.')
+        # If the user requested a specific port endpoint, check if it is exposed
+        if endpoint not in port_details:
+            error_msg = (f'Port {endpoint} not exposed yet. '
+                         f'{_ENDPOINTS_RETRY_MESSAGE} ')
```
opening ports with `LoadBalancer` service takes ~45s on GKE and upto ~1 min on EKS. Maybe we can let it happen async while SkyPilot completes the rest of `setup` and `run` sections?

romilbhardwaj commented on sky/cli.py:
```diff
@@ -2665,7 +2614,8 @@ def _down_or_stop_clusters(
                     hint_or_raise(controller_name)
-                except exceptions.ClusterOwnerIdentityMismatchError as e:
+                except (exceptions.ClusterOwnerIdentityMismatchError,
+                        RuntimeError) as e:
```
Good point - added a docstr for `_hint_or_raise_for_down_sky_serve_controller` and `_hint_or_raise_for_down_spot_controller`.

romilbhardwaj commented on sky/backends/backend_utils.py:
```diff
@@ -2681,3 +2689,101 @@ def check_stale_runtime_on_remote(returncode: int, stderr: str,
+                raise RuntimeError(error_msg)
+        return {endpoint: port_details[endpoint][0].url()}
+    else:
+        if not port_details:
+            # If cluster had no ports to be exposed
+            if handle.launched_resources.ports is None:
+                with ux_utils.print_exception_no_traceback():
+                    raise ValueError('Cluster does not have any ports '
+                                     'to be exposed.')
+            # Else ports have not been exposed even though they exist.
+            # In this case, ask the user to retry.
+            else:
+                error_msg = (f'No endpoints exposed yet. '
+                             f'{_ENDPOINTS_RETRY_MESSAGE} ')
+                if handle.launched_resources.cloud.is_same_cloud(
+                        clouds.Kubernetes()):
+                    # Add Kubernetes specific debugging info
+                    error_msg += \
+                        kubernetes_utils.get_endpoint_debug_message()
+                with ux_utils.print_exception_no_traceback():
+                    raise RuntimeError(error_msg)
```
I thought about it and we could do it. However, if we keep the port opening [non-blocking](https://github.com/skypilot-org/skypilot/pull/3377#discussion_r1560120112), it might be safer to raise exceptions here instead of returning an empty dict so the caller does not silently get an empty dict if the ports are not open yet. wdyt?

romilbhardwaj replied:
I see - I have updated it to return empty dict instead of raising errors in this commit: https://github.com/skypilot-org/skypilot/pull/3377/commits/a0ba0d17f91ad465dbf1bc43e87e2ecde92b5ba4

One side effect of using logging.warning is that it may be hard to suppress in certain cases and the handling now needs to done in the caller (E.g., in cli.py). Perhaps that's ok?

Here's what the errors now looks like:
```
$ sky status --endpoint 8888 k8s2
W 04-29 13:06:03 backend_utils.py:2771] Port 8888 not exposed yet. If the cluster was recently started, please retry after a while. Additionally, make sure your LoadBalancer is configured correctly.
W 04-29 13:06:03 backend_utils.py:2771] To debug, run: kubectl describe service
Aborted!

$ sky status --endpoints k8s2
W 04-29 13:06:04 backend_utils.py:2791] No endpoints exposed yet. If the cluster was recently started, please retry after a while. Additionally, make sure your LoadBalancer is configured correctly.
W 04-29 13:06:04 backend_utils.py:2791] To debug, run: kubectl describe service
Aborted!
```

wdyt?

romilbhardwaj commented on sky/backends/backend_utils.py:
```diff
@@ -2681,3 +2689,101 @@ def check_stale_runtime_on_remote(returncode: int, stderr: str,
                     f'\n--- Details ---\n{stderr.strip()}\n')
+
+
+def get_endpoints(cluster: str,
```
Good point - I've added `def endpoints()` to `sky.core`. However, I'm letting `serve/core.py` directly call `backend_utils.get_endpoints` because `from sky import core` in `serve/core.py` causes a circular import.

romilbhardwaj commented:
Thanks for the review @Michaelvll! Addressed comments, let me know what you think. If looks good, I'll start running smoke tests.

Michaelvll commented on sky/utils/common_utils.py:
```diff
@@ -61,11 +61,17 @@ def get_usage_run_id() -> str:
+    Args:
+        force_fresh_hash: Bypasses the cached hash in USER_HASH_FILE and the
+            hash in the USER_ID_ENV_VAR and forces a fresh user-machine hash
+            to be generated.
```
It might be good to specify the reason we need this argument in the comment. : )
```suggestion
            to be generated. This is only used for `kubernetes.ssh_key_secret_field_name` to avoid
            controllers sharing the same ssh key field name with the local laptop.
```

Michaelvll reviewed (commented):
Thanks for updating the PR @romilbhardwaj! The PR looks mostly good to me with mostly nits.

romilbhardwaj commented on llm/vicuna/serve.yaml:
```diff
@@ -19,11 +19,12 @@ run: |
   echo 'Starting model worker...'
   python -u -m fastchat.serve.model_worker \
             --model-path lmsys/vicuna-${MODEL_SIZE}b-v1.3 2>&1 \
+            --host 127.0.0.1 \
```
Yes, related to https://github.com/lm-sys/FastChat/issues/1714. `localhost` in a k8s container may resolve to `::1` in environments with ipv6 enabled - more details: https://github.com/docker/for-linux/issues/1374

romilbhardwaj commented on sky/backends/backend_utils.py:
```diff
@@ -2681,3 +2689,104 @@ def check_stale_runtime_on_remote(returncode: int, stderr: str,
+    port_details = provision_lib.query_ports(repr(cloud),
+                                             handle.cluster_name_on_cloud,
+                                             handle.launched_resources.ports,
+                                             head_ip=handle.head_ip,
```
Agreed, but the challenge is to support clouds not on the new provisioner without adding special case handling for each of them..

romilbhardwaj commented on sky/serve/controller.py:
```diff
@@ -152,7 +152,10 @@ def configure_logger():
+        # We expose the controller to the public network to allow external
+        # load balancers (example, for high availability load balancers) to
+        # communicate with the controller.
+        uvicorn.run(self._app, host='0.0.0.0', port=self._port)
```
Adding auth might out of scope for this PR. For now, I have changed the logic to expose on `0.0.0.0` only if running inside kubernetes: https://github.com/skypilot-org/skypilot/pull/3377/commits/803df4ab7dd95f9e7f3834199071c501e6207052

This will be accessible from only within the kubernetes cluster (which is a trusted domain) and is not exposed to the public internet. Does that work?

romilbhardwaj commented:
Thanks for the reviews @Michaelvll! Ready for another look. If it looks ok at a high level, I'll start running smoke tests.

Michaelvll reviewed (approved):
Thank you for the update @romilbhardwaj and sorry for the delay! LGTM. Could you help resolve the conflict? If the tests work fine, I think it is good to go.

Michaelvll commented on docs/source/serving/sky-serve.rst:
```diff
@@ -308,11 +308,12 @@ Let's bring up a real LLM chat service with FastChat + Vicuna. We'll use the `Vi
       conda activate chatbot
 
       echo 'Starting controller...'
-      python -u -m fastchat.serve.controller > ~/controller.log 2>&1 &
```
nit: should we also mention serving on kubernetes a bit in the serving doc and refer to the cloud-permissions for the setup?
We may also want to explicitly list out why people should use skyserve on kubernetes than just using deployments, such as:
1. easier customization for load balancer and autoscaling policy
2. simple interface and easy to setup
3. spillover to clouds.

romilbhardwaj replied:
Good idea - filed https://github.com/skypilot-org/skypilot/issues/3518, will submit a separate PR for docs.

romilbhardwaj commented:
Thanks @Michaelvll! Addressed comments. Running smoke and backcompat tests now:

- [ ] Cloud smoke tests, including serve tests - `pytest -v tests/test_smoke.py`
- [ ] Kubernetes smoke tests, including serve tests - `pytest -v tests/test_smoke.py --kubernetes`
- [ ] Backward compatibility - run a service from master, switch branch and launch a new svc. Try to access both services.

romilbhardwaj commented:
Update - had to add `skip_status_check` arg to `backend_utils.get_endpoints()` to disable checks for cluster status (i.e., do not fail early if cluster is not UP). This is similar to the previous behavior, where `head_ip:port` was used without checking cluster status. This is required when multiple `sky serve up` are run concurrently (e.g., in smoke tests), since each call would put the cluster in `INIT` state and get_endpoints call at the end of `serve.core.up` would fail:

```
  File "/home/gcpuser/.sky/file_mounts/sky_repo/sky/serve/core.py", line 257, in up
    endpoint = backend_utils.get_endpoints(
  File "/home/gcpuser/.sky/file_mounts/sky_repo/sky/backends/backend_utils.py", line 2756, in get_endpoints
    raise exceptions.ClusterNotUpError(
sky.exceptions.ClusterNotUpError: Cluster 'sky-serve-controller-c3f7de47' is not in UP status.
```

Smoke tests pass now, this is ready to be merged.

Michaelvll commented on tests/test_yamls/minimal.yaml:
```diff
@@ -2,6 +2,7 @@ name: min
 
 setup: |
   echo "running setup"
+  sudo apt-get install -y jq
```
Would be nice to move this to `test_minimal` in `test_smoke.py` only. : )
